Restore Fusion index to that of 3.3

burqen opened the pull request:
Disable
- Native string index
- Native Temporal index
- Native Spatial index

Also make Fusion index multiversionable for future adaptation.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/helpers/ArrayUtil.java:
```diff
@@ -431,14 +433,33 @@ public static boolean isEmpty( Object[] array )
         return -1;
     }
 
+    public static int[] without( int[] source, int... toRemove )
```
please use ArrayUtils.removeAll instead

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessor.java:
```diff
@@ -100,7 +102,10 @@ public IndexReader newReader()
-        BoundedIterable<Long>[] entries = instancesAs( BoundedIterable.class, IndexAccessor::newAllEntriesReader );
+        @SuppressWarnings( "unchecked" )
+        BoundedIterable<Long>[] entries = Arrays.stream( instancesAs( BoundedIterable.class, IndexAccessor::newAllEntriesReader ) )
+                .filter( notNull() )
```
i think filtering out and guarding against nulls everywhere is a source of errors, what if instead we will have those instance as an empty instance of some sort?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java:
```diff
@@ -0,0 +1,47 @@
+
+import static org.neo4j.kernel.impl.index.schema.fusion.FusionIndexBase.LUCENE;
+
+public class FusionSelector00 implements FusionIndexProvider.Selector
```
java docs are not here at all right now (in all of the selectors)

MishaDemianenko reviewed (changes requested):
Couple of macro comments, lets talk about those

Commit:
Restore Fusion index to that of 3.3

Disable
- Native string index
- Native Temporal index
- Native Spatial index

Also make Fusion index multiversionable for future adaptation.

burqen commented:
@MishaDemianenko Sorry for force pushing the fixes, but it was the easiest way to revert all of the null checks. This looks a lot better I think. Thanks for the feedback!

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexBase.java:
```diff
@@ -163,4 +166,19 @@
         }
     }
+
+    static void validateSelectorInstances( Object[] instances, int... notNullIndex )
```
thats not really notNullIndex now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:
```diff
@@ -76,103 +71,28 @@ public IndexSampler createSampler()
     {
-        if ( predicates.length > 1 )
+        IndexReader instance = selector.select( instances, predicates );
+        if ( instance != null )
```
thats no longer so right?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:
```diff
@@ -76,103 +71,28 @@ public IndexSampler createSampler()
-
-        if ( predicate instanceof RangePredicate )
+        IndexReader instance = selector.select( instances, predicates );
+        if ( instance != null )
```
and here?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Selector for index provider "lucene-1.x".
+ * The version name "00" comes from lucene-1.x originally not being a fusion index.
+ * todo If spatial and temporal will live together with lucene-1.x, update this selector. Otherwise, remove it.
```
temporal

MishaDemianenko reviewed (changes requested):
Thats indeed much better but still have some questions

Commit:
Small cleanup

tinwelint commented:
Closing and reopening when development has gotten further.
